content/spin/v2/testing-apps.md

```diff
@@ -0,0 +1,99 @@
+title = "Testing Applications"
+template = "spin_main"
+date = "2024-05-05T00:00:01Z"
+[extra]
+url = "https://github.com/fermyon/developer/blob/main/content/spin/v2/testing-apps.md"
+
+---
+
+The `spin-test` plugin allows you to run tests, written in WebAssembly, against a Spin application (where all Spin and WASI APIs are configurable mocks).
+
+To use `spin-test` you write test scenarios for your app in any language, with WebAssembly component support, and mock out all interactions your app has with the outside world without requiring any code changes to the app itself. That means the code you test in development is the same code that runs in production.
+
+## Prerequisites
+
+Spin test requires Spin 2.5 or newer. Please [install](./install.md) or [upgrade](./upgrade.md) Spin to begin.
+
+## Installing the Plugin
+
+To run `spin-test` , you’ll first need to install the canary release of the plugin. As `spin-test` matures, we’ll be making stable releases:
+
+```bash
+spin plugin install -u https://github.com/fermyon/spin-test/releases/download/canary/spin-test.json
+```
+
+This will install the plugin which can be invoked with `spin test`:
+
+## Writing a Test
+
+Next, you'll need a test that `spin-test` can run compiled to a WebAssembly component.
+
+There is currently first-class support for Rust, but any language with support for writing WebAssembly components can be used as long as the `fermyon:spin-test/test` world is targeted. You can find the definition of this world [here](https://github.com/fermyon/spin-test/blob/4dcaf79c10fc29a8da2750bdaa383b5869db1715/host-wit/world.wit#L13-L16).
+
+Here’s an example of a test written in Rust using the [Spin Test Rust SDK](https://github.com/fermyon/spin-test) that tests to ensure that the Spin app responds properly when the key-value store has a certain key already set:
+
+```rust
+use spin_test_sdk::{
```

Comment:
Is there any info on the SDK (how to add to yourn project, how to use in your test code)?  (Similarly for equivalents for other languages.)

Reply:
There is only an SDK for Rust at the current moment. Other languages can be used but they require raw bindings which obviously requires a bit more comfort with component model tooling.

As for more info on the Rust SDK, there isn't currently any more. As a start we can expand on this example with the following points:
* Each function marked with `#[spin_test]` will be run as a separate test
* Each test can perform any setup to their environment by using the APIs available in `spin_test_sdk::bindings::fermyon::spin_test_virt`
* To make a request to the Spin application use the `spin_test_sdk::perform_request` function
* After requests are made, you can use the APIs in `spin_test_sdk::bindings::fermyon::spin_test_virt` to check how the request has changed the state of various resources (such as the key/value store) and make assertions that things changed in the way you expected.

This is also all very new and unpolished so more than happy to make changes to `spin_test` itself to make things easier to document/teach.

Reply:
Thanks, @rylev. This is super helpful.